### ISSUE TITLE
Incoming requests (WSGI) is not distributed in a round robin fashion between workers

### ISSUE THREAD

**Issue by moorebill** [eventlet/eventlet#635]
We have a very high I/O bound tasks (disk writes), which made us to have a large set of workers (forked and monkey patched), with a very low pool size to distribute the requests among those workers. The requests which are coming is being pushed towards a particular worker, instead of round robin distribution. Is there a way to accomplish a round robin for the incoming requests?

**Comment by temoto** [comment]
In general, I'd recommend something like https://mongrel2.org/ so free worker can pull request from queue instead of some never-enough-smart front-proxy decides to push a request into already busy worker.

But if that's not an option, please show code, because I don't really understand what's going on.

**Comment by moorebill** [comment]
https://opendev.org/openstack/swift/src/branch/master/swift/common/wsgi.py#L555

There are requests coming in to wsgi server, instead of a round robin, it places (mostly adhoc) on a already oversubscribed worker's queue to process.

**Comment by temoto** [comment]
@moorebill ok so there is nothing interesting in that code, it's very straight forward, good.

- how OS schedules connections to particular process, is kind of out of scope, let's hold this point until more info is available
- so we really should focus debug on single process scenario, anything more should just work better
- eventlet.wsgi just calls `pool.spawn()`
- you use custom pool, but it's irrelevant since `spawn` method is not changed
- GreenPool is not like connection pool, it's just a wrapper on top of a semaphore (counter)
- when GreenPool is under capacity, it just `eventlet.spawn()`-s a new greenthread, there is no persistent worker; GreenPool capacity is kind of queue; i can only hope it makes sense

So are you not satisfied with wsgi scheduler (which is just a semaphore in this regard), or some other disk writing pool? Because disk write blocks everything, including ability to switch greenthreads, you should do blocking operations in multiprocessing or something similar.

**Comment by tipabu** [comment]
FWIW, Swift is working on a patch that should hopefully improve the situation: https://review.opendev.org/#/c/745603/

TL;DR: since eventlet's `listen` already uses SO_REUSEPORT, Swift should just need to open a fresh listen socket per worker process to improve the OS scheduling.

@moorebill this might be better suited to comments on the Swift patch or in IRC at #openstack-swift on freenode, but are you seeing this more with the object-server or proxy-server? How many workers do you have running for your servers? Are you running with `servers_per_port`, which allows you to have multiple worker processes per disk? Can you try out the patch and see if you observe improvements? I'd love feedback on the patch if you have a chance!

**Comment by moorebill** [comment]
This one is container service. SQLITE DB gets so huge, like 300G, and any update/insert/delete can block the greenthread, there by starving others. To avoid this starvation the plan was to flatten the requests, by having n workers (1024) and just greenpool size of 32. Even though 50% of these workers are sitting ideal, the rest are maxed out.

**Comment by temoto** [comment]
@moorebill listen to Tim, obviously he knows stuff related to Openstack. But IMHO this is a perfect use case for `eventlet.tpool`. Full recipe:
- one eventlet.wsgi process is enough, increase in case you may also receive CPU intensive requests
- set wsgi pool size to max number of concurrent requests you want to process; this is your backpressure tuning knob 1; excessive requests will wait in Protocol.process_request
- run wsgi with env `EVENTLET_THREADPOOL_SIZE=1024` this is your backpressure tuning knob 2; excessive requests will wait at `tpool.execute`
- with little bit of code you can implement 503 service overloaded in both cases above
- wrap blocking calls with `eventlet.tpool.execute(sqlite_query, arg1, kw=arg)`

I'd start with setting both pool sizes to same number and maybe you can find better combination.

Please share your results. If I understand correctly, so far it doesn't seem a bug in eventlet.

**Comment by moorebill** [comment]
Only needs backporting of the resuseport fix (our old eventlet 0.20.0), from our end. Eventlet newer version has already the fix for it.

**Comment by temoto** [comment]
Also try that RestrictedGreenPool with `spawn = spawn_n` (because wsgi calls spawn) and pool size = 1.

**Comment by tipabu** [comment]
@temoto IDK _that_ much about OpenStack, mainly just Swift :wink:

@moorebill You might also want to look at backporting https://review.opendev.org/#/c/727876/ if it's a container with a lot of churn. If it's just got a lot of objects in it, your best path forward is to upgrade so you can take advantage of container sharding (introduced in rocky, but with many improvements since). FWIW, the database layers (account and container) demand a lot of IO in general -- if they're on spinning disks it'll be _hugely_ beneficial to upgrade to SSDs.

Oh, and if you want to try the tuning suggestions:
* set `workers = 1` to have a single wsgi process
* use `max_clients` to adjust the wsgi pool size

Using a tpool when making the blocking call is a good suggestion. I know we do that in the object-server whenever we're about to touch a disk; I should look at doing it in the database servers, too.

**Comment by sudhakar-datta** [comment]
Sergey's idea of tpool is a very good. Testing it as of now. Have tuned the workers, max_clients and the backlog.

**Comment by moorebill** [comment]
Sergey's idea of tpool is a very good. Testing it as of now. Have tuned the workers, max_clients and the backlog.
